pkg/queue/protobuf_stats_reporter.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"errors"
+	"io"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+
+	"knative.dev/serving/pkg/autoscaler/metrics"
+	"knative.dev/serving/pkg/network"
+)
+
+const (
+	contentEncodingHeader = "Content-Encoding"
+	contentTypeHeader     = "Content-Type"
+)
+
+// PrometheusStatsReporter structure represents a prometheus stats reporter.
+type ProtobufStatsReporter struct {
+	reportingPeriod time.Duration
+	startTime       time.Time
+	stat            atomic.Value
+	podName         string
+}
+
+// NewPrometheusStatsReporter creates a reporter that collects and reports queue metrics.
+func NewProtobufStatsReporter(pod string, reportingPeriod time.Duration) (*ProtobufStatsReporter, error) {
+	if pod == "" {
+		return nil, errors.New("pod must not be empty")
+	}
+
+	return &ProtobufStatsReporter{
+		reportingPeriod: reportingPeriod,
+		startTime:       time.Now(),
+		stat:            atomic.Value{},
+		podName:         pod,
+	}, nil
+}
+
+// Report captures request metrics.
+func (r *ProtobufStatsReporter) Report(stats network.RequestStatsReport) {
+	// Requests per second is a rate over time while concurrency is not.
+	rp := r.reportingPeriod.Seconds()
+	r.stat.Store(metrics.Stat{
+		PodName:                          r.podName,
+		RequestCount:                     stats.RequestCount / rp,
+		ProxiedRequestCount:              stats.ProxiedRequestCount / rp,
+		AverageConcurrentRequests:        stats.AverageConcurrency,
+		AverageProxiedConcurrentRequests: stats.AverageProxiedConcurrency,
+		ProcessUptime:                    time.Since(r.startTime).Seconds(),
+	})
+}
+
+// Handler returns an uninstrumented http.Handler used to serve stats registered by this
+// ProtobufStatsReporter.
+func (r *ProtobufStatsReporter) Handler() http.Handler {
+	return http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+		stat := r.stat.Load()
+		if stat == nil {
+			httpError(rsp, errors.New("no metrics available yet."))
+		} else {
+			w := io.Writer(rsp)
+			header := rsp.Header()
+			header.Set(contentTypeHeader, "application/protobuf")
+			data := stat.(metrics.Stat)
+			buffer, err := proto.Marshal(&data)
+			if err != nil {
+				httpError(rsp, err)
+			}
+			_, err = w.Write(buffer)
+			if err != nil {
+				httpError(rsp, err)
+			}
```

Comment:
I believe golang servers generally don't handle the write error here as this will always mean that the connections broken anyway AFAIK. The http error is likely never received. I'd drop the error handling.

Reply:
I had a second thought on this yes most of the time it may not even be feasible to talk to client if things fail at this point but usually logging is something that could be proven useful.

Reply:
Sure, if it's not already logged somewhere anyway.